text/3802-assume-bound.md

```diff
@@ -0,0 +1,92 @@
+- Feature Name: `assume_bound`
+- Start Date: 2025-04-21
+- RFC PR: [rust-lang/rfcs#3802](https://github.com/rust-lang/rfcs/pull/3802)
+- Rust Issue: N/A
+
+# Summary
+[summary]: #summary
+
+This feature allows to assume trait bounds on generics so that the caller don't has to proof them pre-monomorph.
+
+# Motivation
+[motivation]: #motivation
+
+I am currently writing a lot of generic magic again.
+To be more concrete, I am making a framework where lots of functions with generic parameters call eachother.
+There, I pass some `impl Key` values, which hold an assoicated type and proof that the generic, which is passed down, implements `Has<T>`, where T is that associated bound.
+Now I don't really want to specify that requirement on every function, as that is cumbersome and hard to maintain. 
+Thus I propose to let me have my unsafe fun to assume trait bounds are fulfilled in the places I am sure about.
+
+Another use case is for complex higher ranked bounds, where a small helper function could be used to hint the compiler that a trait is really implemented.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When implementing a function with a `where` clause, like this one:
+```rs
+pub fn print<T>(val: T)
+where
+    T: Debug
+{ .. }
+```
+and you want to call that from a less restricted generic
+```rs
+fn less_restricted<T>(val: T) {
+    print(val); // error[E0277]: `T` doesn't implement `Debug`
+}
+```
+may it be from a library where you can't restrict it further but know that you only pass in correct types, you could assume the `T: Debug` condition on the upper function:
+```rs
+pub fn print<T>(val: T)
+where
+    #[unsafe(assume)] T: Debug
+{ .. }
+```
+This makes `T` still behave like having `Debug` in the function body, but that isn't the case for the caller. \
+There, it skips the check and just assumes the condition is true.
+
+You may of course not want to change the `print` function, so you could also make a small util function just for the `less_restricted` call, like so:
+```rs
+fn less_restricted<T>(val: T) {
+    fn print_assumed<T>(val: T)
+    where
+        #[unsafe(assume)] T: Debug
+    {
+        print(val);
+    }
+
+    print_assumed(val); // works fine
+}
+```
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+`assume`d bounds are just skipped during bounds check and we trust the user.
```

Comment:
Does this work?

```rust
unsafe fn foo<T>(param: T) -> impl Debug 
where
    #[unsafe(assume)] T: Debug,
{
    param
}
```

Reply:
Yes, it should. Is this something I should explicitly provide as an example?